### PR TITLE
Fix DynamicGroup Serialization Bug

### PR DESCRIPTION
**Pull request by bryanculver**
# Closes: #DNE
# What's Changed

Worked with @lampwins to reproduce this bug.

TLDR: When a Dynamic Group is being saved, there is an edge case where a Filter Form and a _FilterSet_ Form are not identical. A way to reproduce this is with: https://github.com/nautobot/nautobot/tree/u/bryanculver-dg-bug and creating the necessary relationship to filter on it. What happened is the nested filter through the relationship and ultimately is filtering on the `name` of an FK, the FilterSet Form expects inputs to be a list of strings (it's a MultiValueCharFilter after all). But for UX we want to have a selection of names of object instances. We can do this by using the `to_field_name` and processing accordingly. We have logic in Dynamic Groups that does this if the FilterSetForm filed is of the right type. The FilterForm is helping us to this today, but the FilterSetForms need some work to be returning/use the smarter Form field types.

This fix allows to use the FilterForm field first if it's available and falls back if it's not.

Special handling had to be done for Tags because our Tag filter form field isn't a ModelMultipleChoiceField; the FilterSetForm field before was using a ModelMultipleChoiceField.

I'm open to ideas on how to write a test for this which doesn't require a bunch of complex setup like in the branch referenced above.

# TODO

- [x] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/models/groups.py [hunk 1]
```diff
@@ -354,6 +355,9 @@ def set_filter(self, form_data):
         # Use the auto-generated filterset form perform creation of the filter dictionary.
         filterset_form = filterset.form
 
+        # Get the declared form for any overloaded form field definitions.
+        declared_form = get_form_for_model(filterset._meta.model, form_prefix="Filter")
```
Comment (glennmatthews) [inline nautobot/extras/models/groups.py:359]: Do we need to fallback to `filterset_form` if `get_form_for_model` fails?

File: nautobot/extras/models/groups.py [hunk 2]
```diff
@@ -354,6 +355,9 @@ def set_filter(self, form_data):
...
+
         # It's expected that the incoming data has already been cleaned by a form. This `is_valid()`
         # call is primarily to reduce the fields down to be able to work with the `cleaned_data` from the
         # filterset form, but will also catch errors in case a user-created dict is provided instead.
```
Comment (glennmatthews) [inline nautobot/extras/models/groups.py:363]: Should we check `declared_form.is_valid` instead of `filterset_form.is_valid` below?

File: nautobot/extras/models/groups.py [hunk 3]
```diff
@@ -364,10 +368,10 @@ def set_filter(self, form_data):
         # empty/null value fields.
         new_filter = {}
         for field_name in filter_fields:
-            field = filterset_form.fields[field_name]
+            field = declared_form.declared_fields.get(field_name, filterset_form.fields[field_name])
             field_value = filterset_form.cleaned_data[field_name]
 
-            if isinstance(field, forms.ModelMultipleChoiceField):
+            if isinstance(field, (forms.ModelMultipleChoiceField, TagFilterField)):
```
Comment (glennmatthews) [inline nautobot/extras/models/groups.py:374]: What's special about TagFilterField? Perhaps add a comment?

File: nautobot/utilities/forms/fields.py [hunk 4]
```diff
@@ -454,6 +454,8 @@ def get_choices():
         # Choices are fetched each time the form is initialized
         super().__init__(label="Tags", choices=get_choices, required=False, *args, **kwargs)
 
+    to_field_name = "slug"
```
Comment (glennmatthews) [inline nautobot/utilities/forms/fields.py:457]: What'll we need to do in 2.0 to avoid reintroducing the bug, given that tags don't have slugs any more?

**Review by glennmatthews — COMMENTED** [review]
Would it make sense to add a unit/integration test to check for this bug?

File: nautobot/extras/models/groups.py (hunk 1 above)
Comment (bryanculver) [reply nautobot/extras/models/groups.py:359]: We do when we lookup the field below.

File: nautobot/extras/models/groups.py (hunk 2 above)
Comment (bryanculver) [reply nautobot/extras/models/groups.py:363]: We lean against form validation when filtering AFAICT, and ultimately the FilterSetForm should be doing the enforcement since it's what is based on the FilterSet.

File: nautobot/utilities/forms/fields.py (hunk 4 above)
Comment (bryanculver) [reply nautobot/utilities/forms/fields.py:457]: IIRC `name` since that's the unique field in 2.0. This keeps things unbreaking.

**Review by glennmatthews — APPROVED** [review]
Needs change fragment, otherwise looks solid.

**Review by timizuoebideri1 — APPROVED** [review]
